File: Setup/python_sample/sample04_bessel.py
import numpy as np      # 科学技術用パッケージの準備

from pyautd3 import (   # AUTD3パッケージの準備
    AUTD3,                  # 触覚ボードの定義
    Controller,             # 触覚ボード制御
    Bessel,                 # ベッセルビームを生成するGain
    BesselOption,           # ベッセルビーム用オプション
    rad,                    # ラジアン単位指定
    Silencer,               # 可聴音抑止
    Sine,                   # 正弦波
    SineOption,             # 正弦波オプション
    Hz,                     # 周波数指定
)
from pyautd3_link_ethercrab import EtherCrab, EtherCrabOption, Status # EtherCrab用パッケージの準備

# デバイスエラー発生時の処理
def err_handler(idx: int, status: Status) -> None:
    # エラー内容の表示
    print( f"Device[{idx}]: {status}" ) 


# デモ開始
print( "<ベッセルビームデモ>" )

# デバイスへの接続(EtherCrabを使用)
autd = Controller.open(
            [ # AUTD3ボードの指定
                AUTD3(pos=[0.0, 0.0, 0.0], rot=[1, 0, 0, 0]) # １枚目のボードの位置と向き
            ],
            EtherCrab( # EtherCrabでの接続
                err_handler=err_handler, # エラー発生時の処理関数指定
                option=EtherCrabOption() # デフォルトオプション
            ) 
        )

# 可聴音を抑えるための指示
autd.send( Silencer() )

# 動作指定
g = Bessel( # ベッセルビームの位相制御
        apex=autd.center(), 
        direction=np.array([0.0, 0.0, 1.0]), 
        theta=13.0 / 180 * np.pi * rad, 
        option=BesselOption()
    )
m = Sine( # Sin振幅
        freq=150 * Hz, # 150Hz
        option=SineOption(), # デフォルトオプション
    )
print( "動作開始" )
autd.send( (m,g) ) # 位相制御と振幅制御を指示

# キー入力待ち(終了指示待ち)
print( "Enterキー入力で終了" )
_= input()

# デバイスへの接続終了
autd.close()

File: Setup/rust_sample/Sample04_bessel.rs
// AUTD3ライブラリの使用宣言
use autd3::prelude::*;

// EtherCrabライブラリの使用宣言
use autd3_link_ethercrab::{EtherCrab, EtherCrabOption};

// ------------------------------------------------------------
// メイン関数
fn main() ->Result<(), Box<dyn std::error::Error>> {

    println!( "<ベッセルビームデモ>" );

	// デバイスへの接続
	println!( "デバイス接続" );
	let devices = [AUTD3 {
			pos: Point3::origin(),
			rot: UnitQuaternion::identity(),
		}];
	let link = EtherCrab::new(
			|idx, status| {
				eprintln!( "Device[{}]: {}", idx, status );
			},
			EtherCrabOption::default()
		); 
	let mut autd = Controller::open( devices, link )?;

	// 可聴音を抑えるための指示
	autd.send( Silencer::default() )?;

	// ベッセルビームの位相制御指示を生成
	let g = Bessel {
				apex: autd.center(),
				dir: UnitVector3::new_normalize( Vector3::new(0.0, 0.0, 1.0) ),
				theta: 13.0 / 180.0 * std::f32::consts::PI * rad,
				option: BesselOption {
				    intensity: Intensity::MAX, 
				    phase_offset: Phase::ZERO,
				}
			};

	// Sin波形の変調制御指示を生成
	let m = Sine {
				freq: 150 * Hz,
				option: SineOption::default()
			};

	// 位相制御と変調制御を送信(動作開始)
    println!( "動作開始" );
    autd.send( (m, g) )?;

	// キー入力待ち(終了指示待ち)
	let mut _s = String::new();
	println!( "Enterキー入力で終了" );
	std::io::stdin().read_line( &mut _s )?;

	// デバイスとの接続終了
	autd.close()?;

	// 終了
	Ok( () )
}

